ZNG spec cleanup (#897)
brimsec/zq: zng/docs/spec.md

> zq attempts to implement everything herein excepting:
>
> * the bytes type is not yet implemented,

> > * ordering hints are not generated or taken advantage of,

Author: I've realized late in the PR review process that this was a dangling reference, as mention of ordering hints was removed from the spec in April, 2020 as part of #508. There's plenty of chatter on Slack around April 3, 2020 among the Dev team all agreeing that it was appropriate to remove ordering hints from the spec since it had never been implemented and there were apparently no plans to implement it, so it seems the dangling reference should go as well.

> Also, we are contemplating reducing the number of primitive types, e.g.,
> the number of variations in integer types.

> * [1. Summary](#1-summary)

Reviewer: 👍

Reviewer: (for the addition of the ToC)

* [2. The ZNG data model](#2-the-zng-data-model)
* [3. ZNG Binary Format (ZNG)](#3-zng-binary-format-zng)
  + [3.1 Control Messages](#31-control-messages)
    - [3.1.1 Typedefs](#311-typedefs)
      - [3.1.1.1 Record Typedef](#3111-record-typedef)
      - [3.1.1.2 Array Typedef](#3112-array-typedef)
      - [3.1.1.3 Set Typedef](#3113-set-typedef)
      - [3.1.1.4 Union Typedef](#3114-union-typedef)
      - [3.1.1.5 Alias Typedef](#3115-alias-typedef)

>     - [3.1.2 End-of-stream Markers](#312-end-of-stream-markers)

Reviewer 2: ```suggestion
    - [3.1.2 End-Of-Stream Markers](#312-end-of-stream-markers)
```

Author: @nwt: According to https://www.grammarcheck.net/capitalization-in-titles-101/:

> Never capitalize prepositions and conjunctions of four or fewer letters

So I've committed a change to capitalize "Stream" but not "of".

Reviewer 2: @philrz: Good catch!

  + [3.2 ZNG Value Messages](#32-zng-value-messages)
* [4. ZNG Text Format (TZNG)](#4-zng-text-format-tzng)
  + [4.1 ZNG Control Messages](#41-zng-control-messages)
    - [4.1.1 Type Binding](#411-type-binding)
    - [4.1.2 Type Alias](#412-type-alias)

>     - [4.1.3 Application-specific Payload](#413-application-specific-payload)

Reviewer 2: ```suggestion
    - [4.1.3 Application-Specific Payload](#413-application-specific-payload)
```

  + [4.2 Type Grammar](#42-type-grammar)
  + [4.3 ZNG Values](#43-zng-values)
    - [4.3.1 Character Escape Rules](#431-character-escape-rules)
    - [4.3.2 Value Syntax](#432-value-syntax)
  + [4.4 Examples](#44-examples)
* [5. Related Links](#5-related-links)

> ## 1. Summary

Reviewer 2: s/Summary/Introduction/

Author: 🤷‍♂️  Works for me. I actually did waffle between "Summary" and "Introduction" and didn't feel strongly either way, so I flipped the coin and went with "Summary". Since I have at least one tiebreaker vote now, I've flipped it to "Introduction" per your suggestion.

```
Here, the type ID is the integer "9" representing the string type
(defined in [Typedefs](#typedefs)) and the data value "hello, world"

> is an instance of a string.

Author: 👍

The `<nfields>` is encoded as a `uvarint`.

The field name is encoded as a UTF-8 string defining a "ZNG identifier".
The UTF-8 string

> is further encoded as a "counted string", which is a `uvarint` encoding

Reviewer 2: ```suggestion
is further encoded as a "counted string", which is the `uvarint` encoding
```

Author: 👍

is encoded as a `uvarint` representing the length of the name in bytes,
followed by that many bytes of UTF-8 string.

It is an error to define an alias that has the same name as a primitive type.
It is also an error to redefine a previously defined alias with a

> type that differs from the original definition.

Author: I moved this up from the TZNG section, as it reads as if it's about ZNG in general.

is encoded as a `uvarint` representing the length of the name in bytes,
followed by that many bytes of UTF-8 string.

> ### 3.1.2 End-of-stream Markers

Reviewer 2: ```suggestion
### 3.1.2 End-Of-Stream Markers
```

be re-emitted
(and note that the typedef may now be assigned a different ID).

> ### 3.2 Value Messages

Author: Maybe I went a little crazy with the scissors here, but while I was cleaning up the TZNG section, I realized we were being inconsistent with the headers, i.e. redundant references to "ZNG" in headers of sub-sections that had already been described in a higher-level header as being about ZNG. Rather than have lots of redundant ZNG refs in the ZNG headers & redundant TZNG refs in the TZNG section, I opted for consistent brevity.

------------------------
```

It is an error for a value to reference a type ID that has not been previously

> defined by a typedef scoped to the stream in which the value appears.

Author: I moved this up from the TZNG section, as it reads as if it's about ZNG in general.

| `bstring`  | variable |  UTF-8 byte sequence with `\x` escapes           |
| `enum `    | variable |  UTF-8 bytes of enum string                      |
| `ip`       | 4 or 16  |  4 or 16 bytes of IP address                     |

> | `port`     | 2      |  2 byte unsigned int          |

Reviewer 2: Nits:
1. `s/2 byte unsigned int/unsigned int of length 2/` for consistency with the other integer types.
2. Add spaces to keep the table pretty.

Author: 1. Strictly speaking, the `...of length...` notation seems to be used for the ones that list `variable` in the `N` column, whereas most of the ones that have a static number in the `N` column lead off with a reference to that static number like @alfred-landrum's suggestion was doing, albeit with a slightly different wording, i.e. if I was trying to follow that pattern I might have landed on `2 bytes of unsigned int`. But that reads a little odd to me, so I've just gone with your suggestion.

2. You're certainly right about the spaces, so I've fixed that to.

> | `port`     | XXX      |  XXX - to be filled in during PR review          |

Author: I noticed there's an entry for the `port` type in the section 3.1.1 Typedefs table, and it seems to be in the code, but it was't mentioned in this table here. I'm hoping someone who knows the code could plug in the correct suggestion. @alfred-landrum sounded like he's game to dig this up.

> | `port`     | XXX      |  XXX - to be filled in during PR review          |

Reviewer 3: ```suggestion
| `port`     | 2      |  2 byte unsigned int          |
```

Reviewer 2: Nit: `s/2 byte unsigned int/unsigned int of length 2/` for consistency with other rows.

A line that begins with `#` is a control message and all other lines
are values.

> ### 4.1 ZNG Control Messages

Reviewer: ```suggestion
### 4.1 TZNG Control Messages
```

Author: @henridf (and cc: @nwt since he gave a 👍  to this comment): Good catch. Upon closer reading though, it seems like there may be more to do here than one surgical change. Throughout the TZNG section, there's places where the spec could probably stand to be explicit in clarifying where we're talking specifically about TZNG vs. general characteristics of ZNG that apply to both types of presentation. This comes up early in the intro text of Section 4 on TZNG: _"A ZNG file/stream is encoded with UTF-8"_ should probably say _"TZNG"_, for instance. If you guys agree, I'll take a shot at making the correct changes throughout that section and ask for a re-review on it.

Author: @henridf: I've made the changes to clean up the TZNG section, so PTAL. Not only did I mention TZNG explicitly now in several places, it also led me to make some of the headers more consistent and notice a couple bits that probably belonged in the ZNG section, so I moved those up. I've added new PR comments to call attention to where I did that. I also noticed a couple references to "ZNG type syntax" that I realized should be a link to somewhere, but it's not obvious where, so I'm hoping someone can give me a tip there. That someone could be you. 😄

Reviewer: (https://github.com/brimsec/zq/pull/897/files/6d6554fc93539db4e8ee8b4ad87cfc144b465c16#r439974371)

A line that begins with `#` is a control message and all other lines
are values.

### 4.1 Control Messages

TZNG control messages have one of four forms defined below.

Any line beginning with `#` that does not conform with the syntax described here
is an error.
When errors are encountered parsing TZNG, an implementation should return a
corresponding error and allow TZNG parsing to proceed if desired.

### 4.1.1 Type Binding

A TZNG type binding has the following form:
```
#<type-tag>:<type-string>
```
Here, `<type-tag>` is a string decimal integer and `<type-string>`

> is a string defining a type according to the ZNG type syntax. They create

Author: Can someone help point to the right place in the spec for the "ZNG type syntax" it's referencing here? I should surely put a hyperlink here.

Reviewer: This is referring to section 4.2 below.

Author: Thanks @henridf. Based on that, I've not only added a hyperlink but set the link text to "TZNG type grammar" so that way a reader familiar with a spec will know what it's referencing without even having to click on the link.

a binding between the indicated tag and the indicated type.

### 4.1.2 Type Alias

A TZNG type alias has the following form:
```
#<type-name>:<type-string>
```
Here, `<type-name>` is an identifier and `<type-string>`

> is a string defining a type according to the ZNG type syntax. They create a

Author: Same as previous comment: Can someone help point to the right place in the spec for the "ZNG type syntax" it's referencing here? I should surely put a hyperlink here.

binding between the indicated tag and the indicated type.
This form defines an alias mapping the identifier to the indicated type.
`<type-name>` is an identifier with semantics as defined in [Section 3.1.1.5](#3115-alias-typedef).

### 4.1.3 Application-Specific Payload

A TZNG application-specific payload has the following form:
```
#!<control-code>:<payload>
```

> Here, `<control-code>` is a decimal integer in the range 6-127 and `<payload>`

Author: I noticed that the majority of our other `<things-in-angle-brackets>` use hyphens in place of spaces, so I've made that consistent here.

When errors are encountered parsing ZNG, an implementation should return a
corresponding error and allow ZNG parsing to proceed if desired.

### 4.1.1 Type Binding

A type binding has the following form:
```
#<type-tag>:<type-string>
```
Here, `<type-tag>` is a string decimal integer and `<type-string>`

> is a string defining a type according to the ZNG type syntax, which creates

Author: The original wording was grammatically odd. After my adjustment, is this text still clearly communicating the original intent?

Reviewer 2: Yes, but it still reads oddly to me. How about this?
```suggestion
is a string defining a type according to the ZNG type syntax. They create
```

Author: 👍

a binding between the indicated tag and the indicated type.

### 4.1.2 Type Alias

A type alias has the following form:
```
#<type-name>:<type-string>
```
Here, `<type-name>` is an identifier and `<type-string>`

> is a string defining a type according to the ZNG type syntax, which creates a

Author: Same as prior comment: Does my `, which` negatively impact what was being communicated here?

Reviewer 2: No, but I think it reads oddly, so how about this?
```suggestion
is a string defining a type according to the ZNG type syntax. They create a
```

Author: 👍

binding between the indicated tag and the indicated type.
This form defines an alias mapping the identifier to the indicated type.
`<type-name>` is an identifier with semantics as defined in [Section 3.1.1.5](#3115-alias-typedef).

It is an error to define an alias that has the same name as a primitive type.
It is also an error to redefine a previously defined alias with a
type that differs from the original definition.

> ### 4.1.3 Application-specific Payload

Reviewer 2: ```suggestion
### 4.1.3 Application-Specific Payload
```

#<type-tag>:<type-string>
```
Here, `<type-tag>` is a string decimal integer and `<type-string>`

> is a string defining a type according to the [TZNG type grammar`](#42-type-grammar). They create

Reviewer: Looks like there's a typo here with a stray backtick after "TZNG type grammar"

Author: Eek. Thanks for noticing. Fixed.

> Here, `<control code>` is a decimal integer in the range 6-127 and `<payload>`

Reviewer: Suggestion: Replace "Here, `<control code>`" with "`Where, `<control code>`"

Author: Actually, if I made that change, I'd think we should drop the comma as well. @nwt is usually good for a tiebreaker in nit situations. Any opinion @nwt?

Reviewer 2: "Where A is B and C is D." doesn't make sense to me.

Reviewer: Oops I meant without comma.

Author: @nwt: Is that your way of saying you prefer the original wording?

Reviewer 2: @philrz: It is, but only because the suggested change appears to turn the sentence into a sentence fragment.

Author: @nwt: Agreed. Based on that I think I'll decline @henridf's suggestion here. Thanks to both for weighing in. 😄

is any UTF-8 string with escaped newlines.

> ### 4.2 Type Grammar

Author: I don't know if it was intentional or an oversight that this section didn't have a number before. However, once there's a TOC, it's usually populated with the headings that make up the sections, and once we start numbering sections it feels odds to have non-numbered, heading-labeled sections, so it seemed appropriate to add one here.

Given the above textual definitions and the underlying ZNG specification, a
grammar describing the textual type encodings is:
```
<stype> := bool | byte | int16 | uint16 | int32 | uint32 | int64 | uint64 | float64

>          | string | bytes | bstring | enum | ip | port | net | time | duration | null

Author: I noticed there's an entry for the `port` type in the section 3.1.1 Typedefs table, and it seems to be in the code, but it was't mentioned in this table here. I'm hoping someone who knows the code could plug in the correct suggestion. @alfred-landrum sounded like he's game to dig this up.

         | <alias-name>

> XXX - I don't see ctype referenced anywhere else. Shouldn't it be?

Author: I'm not a pro at these pseudo-grammars, but just following the pattern of what I see in the ones in this spec, it seems like we only define something in `<brackets>` if it's actually referenced elsewhere, and this one isn't referenced elsewhere, so I want to check if something needs to be corrected here.

Reviewer: Good catch. We could just add:

```
<type-string> := <ctype>
```

to say that a type definition can be any container type. The caveat is that this isn't currently matched by the tzng implementation, which requires that a type definition be a record type. 

This could be ok based on a similar precedent of the zng spec being broader than the current implementation. If however we want the grammar to match the implementation, then we could have `<type-string> := <rtype>` and 

```
<rtype> :=  record [ <columns> ]
               | record [ ]

<ctype> :=  array [ <stype> ]
          | union [ <stype-list> ]
          | set [ <stype> ]
          | <rtype>
```

Reviewer: I pushed a commit to do the above, and also updates `<column>` to consist of both simple and container types. (having realized as I did it that my comment above was incomplete)

Author: Cool, thanks @henridf! While looking at your commit, I noticed for the first time some excess space characters that were in there from before, so I've pushed yet another commit to improve that.

<alias-name> := <id>

<id> := <id-start> <id-continue>*

<id-start> := [A-Za-z_$]

> <id-continue> := <id-start> | [0-9]

Author: I noticed that the majority of our other `<things-in-angle-brackets>` use hyphens instead of underscores, so I've made that consistent here.

> A reference implementation of this type system is embedded in

Reviewer: This comment seems a little out of place just after the TZNG pseudo-grammar. Maybe it should go at the end of section 3?

Author: @henridf: Good catch. I agree it seems out of place. I'm not sure if it belongs at the end of section 3 either. Here's some thinking out loud...

I looked back at the history of this. As I suspected, the line has been there since [the first version of the ZSON spec](https://github.com/brimsec/zq/blob/5fc47f71c7f9b130c3bcd4a2d377735349fd39a9/pkg/zson/docs/spec.md) when there was no binary format yet and it was trying to be backward compatible with Zeek, and hence was pointing at what was then `zq/pkg/zeek` as the reference implementation, since at the time we were implementing the Zeek data types in their entirety. It seems as if the line has lived on and morphed we moved packages around and cut the ties on 100% backward compatibility with Zeek.

Now, even without me fully understanding the code, no doubt that TZNG psuedo-grammar is implemented _somewhere_ in zq. What I can't say is precisely where, or if it's spread out, if the line is helpful as currently written, as it seems to hand-wave at a point in the code tree that implements lots of other things too.

Up-leveling a bit, wouldn't it be fair to say that zq acts as a reference implementation for _everything_ in the ZNG spec? That is, not just the implementation of the type systems in both TZNG and binary ZNG, but also the reader behaviors, error handling, etc. The opening disclaimer at the top of the spec effectively already says that zq is this reference implementation. It seems one thing we could do is remove the line from this spot below the TZNG pseudo-grammar, make it more generic, and move it to the bottom of the Introduction. After all, it seems a fitting way to close that paragraph would be to remind the reader that the spec is backed up by running code.

... or... if we think it's helpful to the reader to give _precise_ pointers to places within the zq source code to where certain parts of the spec are implemented, we could do that, starting by perhaps refining this one to point specifically to where the TZNG grammar is implemented, then for consistency, adding the same at other places in the spec. This feels like a lot of work that I'd need to lean on you Dev guys for, though. Considering it remains to be seen how many people are scrutinizing the spec at that level of detail, I'm _not_ personally advocating that we make that investment. If people start coming to us with such questions on Slack we could answer them a la carte until it becomes such a nuisance that we add this to the spec.

I'm open to other suggestions as well.

Reviewer: I agree with making the sentence more generic and moving it to the intro. And also that doing detailed references to specific parts of the implementation isn't worthwhile for the time being.

Author: @henridf: I've made it generic and moved it to the bottom of the intro. PTAL and let me know what you think. Thanks!

Reviewer: 👍

Reviewer: 👍

`bstring` | a UTF-8 string with `\x` escapes of non-UTF binary data
`enum` | a string representing an enumeration value defined outside the scope of ZNG
`ip` | a string representing an IP address in [IPv4 or IPv6 format](https://tools.ietf.org/html/draft-main-ipaddr-text-rep-02#section-3)

> `port` | a string representing an 16-bit network port |

Reviewer 2: Nit: `s/decimal string representation of an unsigned, 16-bit integer/` for consistency with other integer types.

Author: Fine point. While I was thinking about consistency, I noticed another entry in the same table led with `a decimal` so I shortened that one to just `decimal` to be like the others. Then I noticed the whole table was lacking in spaces to make it pretty at the Markdown level, so I tidied that up a bit while I was at it.

> `port` | XXX - to be filled in during PR review          |

Author: I noticed there's an entry for the `port` type in the section 3.1.1 Typedefs table, and it seems to be in the code, but it wasn't mentioned in this table here. I'm hoping someone who knows the code could plug in the correct suggestion. @alfred-landrum sounded like he's game to dig this up.

> `port` | XXX - to be filled in during PR review          |

Reviewer 3: ```suggestion
`port` | a string representing an 16-bit network port |
```
